### ISSUE TITLE
Split up DForallT into two different constructors

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#113]
This is the current definition of `DForallT`:

https://github.com/goldfirere/th-desugar/blob/9b9a6f0bdb6becd3f5ae896ddb539069131f7ba0/Language/Haskell/TH/Desugar/AST.hs#L43

This isn't too surprising, since it mirrors the exact structure of `template-haskell`'s `ForallT`. But `ForallT`/`DForallT` are a bit strange, since they combine the distinct concepts of type variable quantification and context arguments. This isn't so bad on its own, since if you want to represent a type that _only_ quantified type variables, you can just leave off the context (e.g., `forall a. a` becomes `DForallT [DPlainTV a] [] (DVarT a)` in `th-desugar`), and similarly if you want to represent a type that only has a context. It's a bit annoying to have to provide an empty list of type variable binders/predicates, but it's not the end of the world.

GHC 8.10 will throw an additional wrinkle into this story, since it introduces visible dependent quantification. A type like `forall a. forall b -> Either a b` would be represented in Template Haskell as:

```haskell
ForallT [PlainTV a] [] $ ForallVisT [PlainTV b]
                       $ ConT ''Either `AppT` VarT a `AppT` VarT b
```

Notice that there's an entirely distinct `ForallVisT` constructor. It's mighty similar to `ForallT`, however, and one might wonder why Template Haskell didn't just combine `ForallT` and `ForallVisT` into a single constructor. Ultimately, Template Haskell's design choice here was motivated by two different factors:

1. `ForallT` bundles a context alongside its type variable binders. It feels somewhat strange to bundle a context with `ForallVisT`, since the type variable binders are visible but the context isn't. TH could have removed the context part and put it into its own constructor, but that would have required a breaking change.
2. In order to distinguish invisible and visible `forall`s, TH would need to bundle some sort of `ForallVisFlag` in the new `ForallT` constructor to tell them apart. But that would require... you guessed it, another breaking change.

In short, changing the design here would involve massive breaking changes for a large chunk of TH clients, so Template Haskell compromised by just introducing a separate `ForallVisT` constructor.

Fortunately, we can strive for a better design in `th-desugar` since backwards compatibility between `th-desugar` versions isn't a top priority. I propose that we split up the current `DForallT` constructor into two different constructors:

```haskell
data DType
  = DForallT DForallVisFlag [DTyVarBndr] DType
  | DCxtT DCxt DType 
  | ...

data DForallVisFlag
  = DForallVis
  | DForallInvis
```

Now visible and invisible `forall`s are unified into a single constructor, and context arguments are confined to a separate constructor.

**Comment by goldfirere** [comment]
I like this direction of travel. But I proposed `DConstrainedT` instead of `DCxtT`. That isn't a "context type"; it's a type with a context. If "contexted" were a thing, that would work, but that word is abhorrent. `DConstrainedT` is a bit long, but I don't see better, currently.

**Comment by RyanGlScott** [comment]
`DConstrainedT` sounds fine to me.

-----

Another issue that I encountered when thinking about how `th-desugar` should accommodate `DForallVisFlag` is the `unravel` function:

https://github.com/goldfirere/th-desugar/blob/9b9a6f0bdb6becd3f5ae896ddb539069131f7ba0/Language/Haskell/TH/Desugar/Core.hs#L1554-L1556

It's unclear to me where visible, dependent quantifiers would go in the result type of `([DTyVarBndr], [DPred], [DType], DType)`. Would they go in the list of `DTyVarBndr`s? Probably not, since those usually represent invisible arguments. Would they go in the list of `DType` arguments? Probably not, since those usually represent things that aren't quantifiers.

In light of this, I'm a bit unclear on how to make `unravel` VDQ-aware. Some possible options are:

1. Turn `unravel`'s return type into `([DTyVarBndr], [DTyVarBndr], [DPred], [DType], DType)`, where the first list of `DTyVarBndr`s represents the invisible quantifiers and the second list represents the visible quantifiers. This would be doable, but things would get weird when you nest nested quantifiers like in this type:

   ```haskell
   T :: forall a -> Int -> forall b -> Either a b
   ```

   Our hypothetical version of `unravel` would turn this into:

   ```haskell
   ([], [PlainTV a, PlainTV b], [], [ConT ''Int], [ConT ''Either `AppT` VarT a `AppT` VarT b])
   ```

   But this completely destroys the fact that the `Int` argument was originally _between_ the `a` and `b` arguments.
2. Craft up some sort of auxiliary data type for function arguments, like this one (design subject to bikeshedding):

   ```haskell
   data DFunArg
     = DFAForall DForallVisFlag DTyVarBndr
     | DFAPred DPred
     | DFAAnon DType
   ```

   Then make `unravel :: DType -> ([DFunArg], DType)`. This would make `unravel` slightly more annoying to use, but it would preserve the original structure of the arguments, even down to the level of nested quantifiers.
3. Something else. (Any other ideas?)

**Comment by goldfirere** [comment]
(2). This direction seems inevitable, for better or worse.

Thanks!

**Comment by RyanGlScott** [comment]
More awkwardness:

1. What happens when we desugar a `ForallT`? Presumably, we will want to chunk up the type variable binders and context into a `DForallT` and `DConstrainedT`, respectively. For instance, it seems obvious that:

   ```haskell
   ForallT [PlainTV a] [ConT ''Show `AppT` VarT a] $ VarT a
   ```

   Should desugar into:

   ```haskell
   DForallT DForallInvis [DPlainTV a] $ DConstrainedT [DConT ''Show `DAppT` DVarT a] $ DVarT a
   ```

   But things become murkier when the list of type variable binders or context is empty. What should `ForallT [PlainTV a] [] (VarT a)` desugar to? There are two possible choices:

   a. `DForallT DForallInvis [DPlainTV a] $ DVarT a` (i.e., `forall a. a`)
   b. `DForallT DForallInvis [DPlainTV a] $ DConstrainedT [] $ DVarT a` (i.e., `forall a. () => a`)

   Unfortunately, both `forall a. a` and `forall a. () => a` have the same representation in the TH AST, so there's no way in general to know which one to pick. For what it's worth, when GHC splices in `ForallT [PlainTV a] [] (VarT a)`, it seems to side more towards option (a):

   ```
   λ> :k $(pure (ForallT [PlainTV (mkName "a")] [] (VarT (mkName "a"))))
   <interactive>:1:3-62: Splicing type
       pure (ForallT [PlainTV (mkName "a")] [] (VarT (mkName "a")))
     ======>
       forall a. a
   ```

   Should `th-desugar` adopt a similar convention?

   Another thing to consider is ```ForallT [] [ConT ''Show `AppT` VarT a] (VarT a)```, since it could conceivably desugar to:

   a. ```DConstrainedT [DConT ''Show `DAppT` DVarT a] $ DVarT a``` (i.e., `Show a => a`)
   b. ```DForallT DForallInvis $ DConstrainedT [DConT ''Show `DAppT` DVarT a] $ DVarT a``` (i.e., `forall. Show a => a`)

   Again, GHC leans towards option (a) here:

   ```
   λ> :k $(pure (ForallT [] [ConT ''Show `AppT` VarT (mkName "a")] (VarT (mkName "a"))))
   <interactive>:1:3-78: Splicing type
       pure
         (ForallT
            [] [ConT ''Show `AppT` VarT (mkName "a")] (VarT (mkName "a")))
     ======>
       Show a => a
   ```

   Finally, we is one more head-scratcher: `ForallT [] [] (ConT ''Int)`. There are _four_ different ways to desugar this:

   a. `DConT ''Int` (i.e., `Int`)
   b. `DForallT DForallInvis [] $ DConT ''Int` (i.e., `forall. Int`)
   c. `DConstrainedT [] $ DConT ''Int` (i.e., `() => Int`)
   d. `DForallT DForallInvis [] $ DConstrainedT [] $ DConT ''Int` (i.e., `forall. () => Int`)

   GHC picks option (a) here:

   ```
   λ> :k $(pure (ForallT [] [] (ConT ''Int)))
   <interactive>:1:3-35: Splicing type
       pure (ForallT [] [] (ConT ''Int)) ======> Int
   ```
2. How should we sweeten `DForallT`s and `DConstrainedT`s? One obvious approach is to turn `DForallT` into `ForallT` with an empty context and to turn `DConstrainedT` into `ForallT` with no type variable binders. There is a downside to this approach, however: consider this TH AST:

   ```haskell
   ForallT [PlainTV a] [ConT ''Show `AppT` VarT a] $ VarT a
   ```

   If we roundtrip this through `desugar`/`sweeten`, then this will come back as something different subtly different:

   ```haskell
   ForallT [PlainTV a] [] $ ForallT [] [ConT ''Show `AppT` VarT a] $ VarT a
   ```

   This turns out to be an issue in practice, since there is a test case that checks for this behavior (see `round_trip_types` in `Test/Splices.hs`). I suppose we could remove these tests—after all, there are already a number of TH AST forms that come out of roundtripping in a slightly different format (e.g., `StarT` -> `DConT ''Type` -> `ConT ''Type`).

   One alternative is to special-case sweetening every occurrence `DForallT _ tvbs (DConstrainedT ctxt ty)` so that it returns a single, compact `ForallT tvbs ctxt ty`. Of course, if we go down this road, we'll have to ask questions about how we should sweeten `DForallT _ [] (DConstrainedT ctxt ty)`, `DForallT _ tvbs (DConstrainedT [] ty)`, and `DForallT _ [] (DConstrainedT [] ty)` as well.

**Comment by goldfirere** [comment]
Thanks for this in-depth analysis of the options and their consequences.

I was originally going to suggest that we make `DForallT` take *one* binder instead of a list. Then, the path forward is obvious. However, this would be wrong, because `foo :: forall a b. blah` brings both `a` and `b` into scope in `foo`'s body, while `foo :: forall a. forall b. blah` brings only `a` into scope. (Another infelicity of `ScopedTypeVariables`.)

But (in my opinion) the best answer comes from the goal that consuming `th-desugar` should be easy. That suggests that missing components of a `ForallT` are dropped during desugaring. This means that `ForallT [] [] blah` desugars identically to `blah`.

Oh, blast. That's wrong, too. `foo :: forall. a -> a` is different than `foo :: a -> a`. If we drop empty lists, then we lose that distinction. (Though, it seems, TH doesn't get this right, either.)

I guess I'm advocating for always desugaring a `ForallT` into a `DForallT`, and optionally with a `DConstrainedT`.

As for sweetening: do whatever is easiest. We don't have a round-trip property; that's not really something th-desugar strives for. But, as with desugaring, we should probably preserve the existence of a `forall`.

**Comment by goldfirere** [comment]
See [GHC#16396](https://ghc.haskell.org/trac/ghc/ticket/16396) about the GHC issues.

**Comment by RyanGlScott** [comment]
If I understand https://github.com/goldfirere/th-desugar/issues/113#issuecomment-469882209 correctly, then we should handle desugaring like so:

* `ForallT tvbs [] ty` should desugar to `DForallT DForallInvis tvbs ty`.
* `ForallT [] ctxt ty` should desguar to `DForallT DForallInvis [] $ DConstrainedT ctxt ty`.
* `ForallT [] [] ty` should desugar to `DForallT DForallInvis [] ty`.
* For all other cases, just straightforwardly desugar `ForallT tvbs ctxt ty` to `DForallT DForallInvis tvbs $ DConstraintedT ctxt ty`.

Where `tvbs` and `ctxt` are non-empty.

Similarly, when sweetening:

* `DForallT DForallInvis tvbs (DConstrainedT ctxt ty)` should sweeten to `ForallT tvbs ctxt ty`.
* `DForallT DForallInvis [] (DConstrainedT ctxt ty)` should sweeten to `ForallT [] ctxt ty`.
* `DForallT DForallInvis tvbs (DConstrainedT [] ty)` should sweeten to `ForallT tvbs [] ty`.
* `DForallT DForallInvis [] (DConstrainedT [] ty)` should sweeten to `ForallT [] [] ty`.
* For all other cases, just straightforwardly sweeten `DForallT DForallInvis tvbs ty` to `ForallT tvbs [] ty` and `DConstrainedT ctxt ty` to `ForallT [] ctxt ty`.

A potential gotcha of this approach is that it `th-desugar` roundtripping will turn `() => a` into `forall. () => a`. One might worry that this could cause problems due to the latter not being a well-scoped, type, but I don't think Template Haskell will ever let you do anything bad with it in practice. For instance, if you try splicing in the TH representation of `forall. () => a` into source code, then GHC will just drop the `forall. () =>` part.

**Comment by goldfirere** [comment]
I agree with your plan.

The "gotcha" is inherent in the current design of TH. My musings in GHC#16396 are a roundabout way of suggesting that we change `ForallT` to be more sensible. But I didn't want to just come out and say it, because that's the wrong way around: my motivation is that the current behavior causes bugs that (I think) are structurally impossible to avoid without changing `ForallT`.
